File: Server/src/services/tools/manage_cloud.py
"""
manage_cloud.py - Cloud Services Tool

Handles:
- Unity Cloud Save
- Remote Config
- Unity Authentication
- Cloud Code
"""

from typing import Any
from mcp.types import TextContent
from .utils import generate_tool_response


def get_tool_definition():
    """Returns the tool definition for manage_cloud."""
    return {
        "name": "manage_cloud",
        "description": """Manage Unity Cloud Services including Cloud Save, Remote Config, and Authentication.

Actions:
- auth_initialize: Initialize Unity Authentication
- auth_sign_in_anonymous: Sign in anonymously
- auth_sign_in_email: Sign in with email/password
- auth_sign_up: Sign up new user
- auth_sign_out: Sign out current user
- auth_get_user: Get current user info
- auth_link_account: Link anonymous to email account
- save_initialize: Initialize Cloud Save
- save_data: Save data to cloud
- load_data: Load data from cloud
- delete_data: Delete cloud data
- list_keys: List all saved keys
- config_initialize: Initialize Remote Config
- config_fetch: Fetch remote config
- config_get_value: Get config value
- config_get_all: Get all config values
- config_set_environment: Set config environment
- cloud_code_run: Run cloud code function
- cloud_code_list: List available cloud functions""",
        "inputSchema": {
            "type": "object",
            "properties": {
                "action": {
                    "type": "string",
                    "description": "Action to perform",
                    "enum": [
                        "auth_initialize",
                        "auth_sign_in_anonymous",
                        "auth_sign_in_email",
                        "auth_sign_up",
                        "auth_sign_out",
                        "auth_get_user",
                        "auth_link_account",
                        "save_initialize",
                        "save_data",
                        "load_data",
                        "delete_data",
                        "list_keys",
                        "config_initialize",
                        "config_fetch",
                        "config_get_value",
                        "config_get_all",
                        "config_set_environment",
                        "cloud_code_run",
                        "cloud_code_list"
                    ]
                },
                "email": {
                    "type": "string",
                    "description": "User email"
                },
                "password": {
                    "type": "string",
                    "description": "User password"
                },
                "saveKey": {
                    "type": "string",
                    "description": "Key for cloud save data"
                },
                "saveData": {
                    "type": "object",
                    "description": "Data to save (JSON object)"
                },
                "saveType": {
                    "type": "string",
                    "description": "Type of save",
                    "enum": ["Default", "Public", "Private"]
                },
                "configKey": {
                    "type": "string",
                    "description": "Remote config key"
                },
                "configEnvironment": {
                    "type": "string",
                    "description": "Remote config environment",
                    "enum": ["development", "staging", "production"]
                },
                "cloudFunctionName": {
                    "type": "string",
                    "description": "Cloud Code function name"
                },
                "cloudFunctionParams": {
                    "type": "object",
                    "description": "Parameters for cloud function"
                },
                "projectId": {
                    "type": "string",
                    "description": "Unity Project ID"
                }
            },
            "required": ["action"]
        }
    }


async def handle_tool_call(arguments: dict[str, Any]) -> list[TextContent]:
    """Handle manage_cloud tool calls."""
    action = arguments.get("action")
    
    if not action:
        return generate_tool_response({
            "success": False,
            "error": "'action' parameter is required"
        })

    params = {"action": action}
    
    param_keys = [
        "email", "password", "saveKey", "saveData", "saveType",
        "configKey", "configEnvironment", "cloudFunctionName",
        "cloudFunctionParams", "projectId"
    ]
    
    for key in param_keys:
        if key in arguments and arguments[key] is not None:
            params[key] = arguments[key]

    return generate_tool_response({
        "tool": "manage_cloud",
        "params": params
    })

File: MCPForUnity/Editor/Tools/ManageCloud.cs
using System;
using MCPForUnity.Editor.Helpers;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEngine;

namespace MCPForUnity.Editor.Tools
{
    /// <summary>
    /// Tool for managing Unity Cloud Services (Authentication, Cloud Save, Remote Config).
    /// </summary>
    [McpForUnityTool("manage_cloud")]
    public static class ManageCloud
    {
        public static object HandleCommand(JObject @params)
        {
            if (@params == null)
                return new ErrorResponse("Parameters cannot be null.");

            string action = ParamCoercion.CoerceString(@params["action"], null)?.ToLowerInvariant();
            if (string.IsNullOrEmpty(action))
                return new ErrorResponse("'action' parameter is required.");

            try
            {
                return action switch
                {
                    "auth_initialize" => AuthInitialize(@params),
                    "auth_sign_in_anonymous" => AuthSignInAnonymous(@params),
                    "auth_sign_in_email" => AuthSignInEmail(@params),
                    "auth_sign_out" => AuthSignOut(@params),
                    "auth_get_user" => AuthGetUser(@params),
                    "save_initialize" => SaveInitialize(@params),
                    "save_data" => SaveData(@params),
                    "load_data" => LoadData(@params),
                    "config_initialize" => ConfigInitialize(@params),
                    "config_fetch" => ConfigFetch(@params),
                    "config_get_value" => ConfigGetValue(@params),
                    _ => new ErrorResponse($"Unknown action: '{action}'.")
                };
            }
            catch (Exception e)
            {
                McpLog.Error($"[ManageCloud] Action '{action}' failed: {e}");
                return new ErrorResponse($"Error: {e.Message}");
            }
        }

        private static object AuthInitialize(JObject @params)
        {
            return new SuccessResponse("Unity Authentication setup.", new
            {
                package = "com.unity.services.authentication",
                steps = new[]
                {
                    "1. Install Unity Services packages",
                    "2. Link project in Project Settings > Services",
                    "3. Initialize with UnityServices.InitializeAsync()",
                    "4. Use AuthenticationService.Instance"
                },
                codeExample = @"await UnityServices.InitializeAsync();
await AuthenticationService.Instance.SignInAnonymouslyAsync();"
            });
        }

        private static object AuthSignInAnonymous(JObject @params)
        {
            return new SuccessResponse("Anonymous sign-in configured.", new
            {
                codeExample = "await AuthenticationService.Instance.SignInAnonymouslyAsync();",
                note = "Creates temporary player account, can be linked later"
            });
        }

        private static object AuthSignInEmail(JObject @params)
        {
            string email = ParamCoercion.CoerceString(@params["email"], "");

            return new SuccessResponse("Email sign-in configured.", new
            {
                email = email,
                codeExample = $"await AuthenticationService.Instance.SignInWithUsernamePasswordAsync(\"{email}\", password);",
                note = "Requires Unity Authentication with username/password enabled"
            });
        }

        private static object AuthSignOut(JObject @params)
        {
            return new SuccessResponse("Sign-out configured.", new
            {
                codeExample = "AuthenticationService.Instance.SignOut();"
            });
        }

        private static object AuthGetUser(JObject @params)
        {
            return new SuccessResponse("Get user info.", new
            {
                codeExample = @"var playerId = AuthenticationService.Instance.PlayerId;
var accessToken = AuthenticationService.Instance.AccessToken;
var isSignedIn = AuthenticationService.Instance.IsSignedIn;"
            });
        }

        private static object SaveInitialize(JObject @params)
        {
            return new SuccessResponse("Cloud Save setup.", new
            {
                package = "com.unity.services.cloudsave",
                codeExample = @"// Save data
await CloudSaveService.Instance.Data.Player.SaveAsync(new Dictionary<string, object> { { ""key"", value } });
// Load data
var data = await CloudSaveService.Instance.Data.Player.LoadAsync(new HashSet<string> { ""key"" });"
            });
        }

        private static object SaveData(JObject @params)
        {
            string key = ParamCoercion.CoerceString(@params["saveKey"], "");
            var data = @params["saveData"];

            return new SuccessResponse("Cloud save configured.", new
            {
                key = key,
                codeExample = $"await CloudSaveService.Instance.Data.Player.SaveAsync(new Dictionary<string, object> {{ {{ \"{key}\", data }} }});"
            });
        }

        private static object LoadData(JObject @params)
        {
            string key = ParamCoercion.CoerceString(@params["saveKey"], "");

            return new SuccessResponse("Cloud load configured.", new
            {
                key = key,
                codeExample = $"var data = await CloudSaveService.Instance.Data.Player.LoadAsync(new HashSet<string> {{ \"{key}\" }});"
            });
        }

        private static object ConfigInitialize(JObject @params)
        {
            return new SuccessResponse("Remote Config setup.", new
            {
                package = "com.unity.services.remoteconfig",
                dashboardUrl = "https://dashboard.unity3d.com/",
                codeExample = @"RemoteConfigService.Instance.FetchCompleted += ApplyRemoteSettings;
await RemoteConfigService.Instance.FetchConfigsAsync(new UserAttributes(), new AppAttributes());"
            });
        }

        private static object ConfigFetch(JObject @params)
        {
            return new SuccessResponse("Remote config fetch.", new
            {
                codeExample = "await RemoteConfigService.Instance.FetchConfigsAsync(userAttributes, appAttributes);"
            });
        }

        private static object ConfigGetValue(JObject @params)
        {
            string key = ParamCoercion.CoerceString(@params["configKey"], "");

            return new SuccessResponse("Get remote config value.", new
            {
                key = key,
                codeExample = $@"var stringValue = RemoteConfigService.Instance.appConfig.GetString(""{key}"");
var intValue = RemoteConfigService.Instance.appConfig.GetInt(""{key}"");
var floatValue = RemoteConfigService.Instance.appConfig.GetFloat(""{key}"");
var boolValue = RemoteConfigService.Instance.appConfig.GetBool(""{key}"");"
            });
        }
    }
}
